core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
       responseCallback(responseError)
     }
 
-    replicaManager.appendRecords(
-      newMetadata.txnTimeoutMs.toLong,
-      TransactionLog.EnforcedRequiredAcks,
-      internalTopicsAllowed = true,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    // note that we need to grab the read lock and double check the metadata and coordinator epoch again
+    // until we have appended to log to avoid the scenario that an emigration and immigration happens after
+    // the check when handling the request
+    inReadLock(stateLock) {
```

Comment:
The problem with that is because we have already check that in `handler`, and re-checking it again in the delayed operation sounds like an overkill to me.

I instead made the changes to only watch on transactional id, not txn partition id, as in the makers queue we already maintain the markers per txn partition so that we can call `cancel` on the transactional ids of that txn partition anyways. WDYT?